### PR TITLE
dvc: init at 0.24.3

### LINKED ISSUE #49438: Request to package dvc
## Issue description

dvc (https://github.com/iterative/dvc) is version control system for "biggish" data and code

It integrates with Git but offers a solution for managing and versioning big data sets that come with code commits.

Related technology (https://github.com/iterative/dvc#related-technologies) are like git annex and git lfs.

If nobody gets around to this, I may work on this later.

### PR DESCRIPTION
**Pull request by CMCDragonkai**
###### Motivation for this change

Adds dvc. Fixes https://github.com/NixOS/nixpkgs/issues/49438

DVC is a version control system for data science that integrates with Git. But doesn't really actually need git.

Since it's a application and not a library, I thought to just put it in `pkgs/applications/version-management/dvc/default.nix`.

Futhermore, I've decided to use `python3Packages` set instead of just `pythonPackages` as we should be using Python 3 for any applications now.

They only have a wheel on the on Pypi, and I've asked them to push up their source distribution to the Pypi as well  https://github.com/iterative/dvc/issues/1509.

In the case that is done, this can be changed to use the source distribution and hopefully run some tests as well.

One thing is that after I built this, and tried to run dvc, I get this at the end:

```
/nix/store/vs6d2fjkl4kb3jb7rwibsd76k9v2n4xy-bash-4.4-p23/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/home/cmcdragonkai/.includes_sh/shell_environment_common.conf: line 31: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
Error: the following arguments are required: COMMAND

Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
usage: dvc [-h] [-q | -v] [-V] COMMAND ...

Data Version Control

optional arguments:
  -h, --help     show this help message and exit
  -q, --quiet    Be quiet.
  -v, --verbose  Be verbose.
  -V, --version  Show program's version.

Available Commands:
  COMMAND        Use dvc COMMAND --help for command-specific help.
    init         Initialize dvc over a directory (should already be a git dir).
    destroy      Destroy dvc. Will remove all project's information, data files and cache.
    add          Add files/directories to dvc.
    remove       Remove outputs of DVC file.
    move         Move output of DVC file.
    unprotect    Unprotect data file/directory.
    run          Generate a stage file from a given command and execute the command.
    repro        Reproduce DVC file. Default file name - 'Dvcfile'.
    pull         Pull data files from the cloud.
    push         Push data files to the cloud.
    fetch        Fetch data files from the cloud.
    status       Show the project status.
    gc           Collect garbage.
    add          Add files/directories to dvc.
    import       Import files from URL.
    config       Get or set config options.
    checkout     Checkout data files from cache.
    remote       Manage set of tracked repositories.
    cache        Manage cache settings.
    metrics      Get metrics from all branches.
    install      Install dvc hooks into the repository.
    root         Relative path to project's directory.
    lock         Lock DVC file.
    unlock       Unlock DVC file.
    pipeline     Manage pipeline.
    daemon       Service daemon.
/nix/store/ydk0mfpvn9smcmn72wc9i20slv1d2b79-python3-3.7.2/bin/python3.7: No module named dvc
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by CMCDragonkai** [comment]
What would be the solution to: `/nix/store/ydk0mfpvn9smcmn72wc9i20slv1d2b79-python3-3.7.2/bin/python3.7: No module named dvc`

**Comment by CMCDragonkai** [comment]
Also I still need to figure out how to setup the remotes: https://github.com/iterative/dvc/blob/master/setup.py#L29-L43

I imagine that we can do these as options on the `default.nix` but suggestions would be good.

**Comment by CMCDragonkai** [comment]
Should it be like:

```
  propagatedBuildInputs = [
    ply
    configparser
    zc_lockfile
    future
    colorama
    configobj
    networkx
    pyyaml
    GitPython
    setuptools
    nanotime
    pyasn1
    schema
    jsonpath_rw
    requests
    grandalf
    asciimatics
    distro
    appdirs
  ]
  ++ lib.optional enableGoogle google_cloud_storage
  ++ lib.optional enableAWS boto3
  ++ lib.optional enableAzure azure-storage-blob
  ++ lib.optional enableSSH paramiko;
```
?

**Comment by CMCDragonkai** [comment]
Given that the cloud remotes are in the `extra_requires` of setup.py. Is there anything special required in the `buildPythonApplication` to make sure the `extra_requires` dependency is in fact installed?

**Comment by CMCDragonkai** [comment]
Theres's a problem here: https://github.com/iterative/dvc/issues/1509#issuecomment-457086163

File: pkgs/applications/version-management/dvc/default.nix [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  version = "0.24.0";
+
+  src = fetchFromGitHub {
+    owner = "iterative";
+    repo = "dvc";
+    rev = version;
+    sha256 = "163lnn8z4ig5dgl2jdkrlqwziyn28shswysq7gpamp8ri0g2ccx3";
+  };
+
+  buildInputs = [ git ];
```
Comment (dotlambda) [inline pkgs/applications/version-management/dvc/default.nix]: From https://github.com/iterative/dvc/issues/1509#issuecomment-457083344, it looks like this should be fixed in GitPython. I have pushed ee82616089ff920792b6c265068902c4fa3d4f7f to master.
```suggestion
```

File: pkgs/applications/version-management/dvc/default.nix [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+, enableAzure ? false
+, enableSSH ? false
+}:
+
+with python3Packages;
+buildPythonApplication rec {
+  pname = "dvc";
+  version = "0.24.0";
+
+  src = fetchFromGitHub {
```
Comment (dotlambda) [inline pkgs/applications/version-management/dvc/default.nix]: Add a comment that there's only a wheel on PyPI.

File: pkgs/applications/version-management/dvc/default.nix [hunk 3]
```diff
@@ -0,0 +1,61 @@
+{ lib
+, python3Packages
+, fetchFromGitHub
+, git
```
Comment (dotlambda) [inline pkgs/applications/version-management/dvc/default.nix]:
```suggestion
```

**Comment by dotlambda** [comment]
> Theres's a problem here: iterative/dvc#1509 (comment)

I don't have any problem executing dvc.

**Comment by CMCDragonkai** [comment]
@dotlambda You don't see this problem:

```
/nix/store/ydk0mfpvn9smcmn72wc9i20slv1d2b79-python3-3.7.2/bin/python3.7: No module named dvc
```

At the end of the dvc command?

**Comment by dotlambda** [comment]
Nope, never saw it while playing around a little with dvc using `nix-review pr 54530`.

**Comment by worldofpeace** [comment]
> Nope, never saw it while playing around a little with dvc using `nix-review pr 54530`.

I can reproduce the message @CMCDragonkai is seeing when executing `dvc` like `result/bin/dvc` but not from within a `nix-shell`, which is probably why you didn't see it with `nix-review`.

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -1345,6 +1345,15 @@ in
 
   duperemove = callPackage ../tools/filesystems/duperemove { };
 
+  dvc = callPackage ../applications/version-management/dvc { };
```
Comment (worldofpeace) [inline pkgs/top-level/all-packages.nix]: Why not use `python3Packages.callPackage`?

Comment (dotlambda) [reply pkgs/top-level/all-packages.nix]: Nope, we don't want to do this because we might have to override a Python package some day.

**Comment by CMCDragonkai** [comment]
This might be the problem: https://github.com/iterative/dvc/issues/1509#issuecomment-457413812

Basically here https://github.com/iterative/dvc/blob/master/dvc/daemon.py#L69-L72

It tries to launch `dvc daemon` as  a subcommand through the Python executable. Like `"sys.executable" -m dvc`. Because it's not critical to the process, things still work, but we end up reporting that error.

Other than the change they are going to make to the source code, what would normally be the solution here on Nixpkgs? Should the package somehow set the `PYTHONHOME` as well so that `-m X` works as a subcommand?

File: pkgs/applications/version-management/dvc/default.nix (hunk 3 above)
Comment (CMCDragonkai) [reply pkgs/applications/version-management/dvc/default.nix]: Done.

File: pkgs/applications/version-management/dvc/default.nix (hunk 2 above)
Comment (CMCDragonkai) [reply pkgs/applications/version-management/dvc/default.nix]: Done.

File: pkgs/applications/version-management/dvc/default.nix (hunk 1 above)
Comment (CMCDragonkai) [reply pkgs/applications/version-management/dvc/default.nix]: Done.

**Review by dotlambda — APPROVED** [review]

**Comment by CMCDragonkai** [comment]
We have this problem: https://github.com/iterative/dvc/issues/1509#issuecomment-458366390

The issue is that `dvc` is Python application that uses the `entry_points` configuration to convert the libray module `dvc` into a callable script. This is usually fine as Nix figures it out and creates `result/bin/dvc` and `result/bin/.dvc-wrapped`.

However the problem is that within `dvc` they try to do a fork+exec on the `dvc` program. They do this here: https://github.com/iterative/dvc/blob/master/dvc/daemon.py#L66-L90

Basically try try to get the Python executable with with `sys.executable` and perform fork and exec of `python -m dvc daemon -q`.

Initially this did not work because within the Nix built package there was no way to find the `dvc` library module. Now that they have done https://github.com/iterative/dvc/pull/1531 they can find it, but the executed subprocess cannot load any of the Python dependencies like `colorama`.

I'm not sure what's the best way to solve this problem.

@FRidh are you able to help here?

**Comment by dotlambda** [comment]
You can simply patch the source to call `$out/bin/dvc daemon`.

**Comment by CMCDragonkai** [comment]
I think I fixed it. Basically patched it out so that it just calls `dvc daemon -q`. Interesting even though I didn't add its own store bin path to the PATH, it still works. Not sure if that's a fluke or designed. However this is annoying as subsequent commits will require fixes of that patch each time. It seems the Git patch is more stringent on its requirements, wouldn't a unified patch be more flexible? I don't know how to use it though.

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (CMCDragonkai) [reply pkgs/top-level/all-packages.nix]: What do you mean?

Comment (dotlambda) [reply pkgs/top-level/all-packages.nix]: I mean you shouldn't use `pythonPackages.callPackage` in `all-packages.nix`.

File: pkgs/applications/version-management/dvc/dvc-daemon.patch [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+     """
+-    cmd = [sys.executable]
+-    if not is_binary():
+-        cmd += ['-m', 'dvc']
+-    cmd += ['daemon', '-q'] + args
+-
+-    env = fix_env()
+-    file_path = os.path.abspath(inspect.stack()[0][1])
+-    env['PYTHONPATH'] = os.path.dirname(os.path.dirname(file_path))
++    cmd = ["dvc", "daemon", "-q"] + args
```
Comment (dotlambda) [inline pkgs/applications/version-management/dvc/dvc-daemon.patch]: Please do make use of `$out`. That way, it will also work as expected if multiple versions of dvc are installed.

Comment (CMCDragonkai) [reply pkgs/applications/version-management/dvc/dvc-daemon.patch]: How do you do that in a patch?

Comment (CMCDragonkai) [reply pkgs/applications/version-management/dvc/dvc-daemon.patch]: Does that work with patch files?

Comment (dotlambda) [reply pkgs/applications/version-management/dvc/dvc-daemon.patch]: Yes, you just have to put a placeholder, e.g. `@dvc@` in the patch.

Comment (dotlambda) [reply pkgs/applications/version-management/dvc/dvc-daemon.patch]: In `postPatch`.

**Comment by CMCDragonkai** [comment]
Feel free to take over @jeromebaum.

**Comment by schlichtanders** [comment]
looking forward to have dvc in nixos

what is the status here?

**Comment by schlichtanders** [comment]
@jeromebaum @CMCDragonkai any update?

**Comment by jeromebaum** [comment]
I haven’t done anything further on this. You’re welcome to take over.
Although the way I understood it this PR is basically unmergeable as it
doesn’t follow the relevant nix norms. If you want dvc in mainline nix
you’ll probably have to either look at the other PR referenced here or
start anew. (I’m not sure which is easier/faster, I’m guessing the first
option?)

**Comment by jeromebaum** [comment]
Oh, sorry I just noticed that I was replying on a different PR than I thought. Please ignore my comment on it being unmergeable. That said, I haven't done anything further on this PR either and I'd assume that @CMCDragonkai is not opposed to you taking it over to bring it to completion.

File: pkgs/applications/version-management/dvc/dvc-daemon.patch (hunk 5 above)
Comment (CMCDragonkai) [reply pkgs/applications/version-management/dvc/dvc-daemon.patch]: Just to clarify. Does this mean we patch, then substitute, or substitute the patch and then apply the patch?

Comment (worldofpeace) [reply pkgs/applications/version-management/dvc/dvc-daemon.patch]: In patch
```
@dvc@ # a placeholder for substituteAll
```

in expression
```
patches = [
  (substituteAll {
    src = ./path_to_patch;
    dvc = "${placeholder "out"}/bin/dvc";
  })
];

postPatch = ''
  substituteInPlace path_to_file_that_was_patched_dvc_placeholder \
    --subst-var-by $out
'';
```

> example https://github.com/NixOS/nixpkgs/blob/master/pkgs/desktops/gnome-3/core/evolution-data-server/default.nix#L27

Comment (Rakesh4G) [reply pkgs/applications/version-management/dvc/dvc-daemon.patch]: Can we do something like this: 
In default.nix :
add
`postPatch = ''
    substituteInPlace dvc/daemon.py --subst-var-by dvc "$out/bin/dcv"
  '';`

in patch file
modify
`cmd = [ *"dvc"* , "daemon", "-q"] + args to cmd = [ *"@dvc@"* , "daemon", "-q"] + args`

Comment (CMCDragonkai) [reply pkgs/applications/version-management/dvc/dvc-daemon.patch]: Seems like it should be easier to just substitute into the patch directly and then apply the patch as suggested by @Rakesh4G? What are your thoughts @worldofpeace @dotlambda

**Comment by CMCDragonkai** [comment]
Just one thing left to do and this can be merged.

File: pkgs/applications/version-management/dvc/dvc-daemon.patch (hunk 5 above)
Comment (worldofpeace) [reply pkgs/applications/version-management/dvc/dvc-daemon.patch]: Either one is fine, I just happen to use the first option we suggested all over nixpkgs.

**Comment by Rakesh4G** [comment]
Hi @CMCDragonkai , I am attaching updated default.nix and patch file for you reference here.

[dvc_nix.txt](https://github.com/NixOS/nixpkgs/files/3760402/dvc_nix.txt)
[dvc_patch.txt](https://github.com/NixOS/nixpkgs/files/3760403/dvc_patch.txt)

**Comment by CMCDragonkai** [comment]
This now works. I've tested a `nix-build` for both targets. Ready to be merged.

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -1345,6 +1345,15 @@ in
...
+
+  dvc-with-remotes = callPackage ../applications/version-management/dvc {
```
Comment (FRidh) [inline pkgs/top-level/all-packages.nix]: Is this variant needed? Users can override themselves, not.

**Comment by CMCDragonkai** [comment]
Its for covenience mainly. I guess each one could itself be a different target. Sort of like vimHugex.

On 28 October 2019 23:19:57 GMT+11:00, Frederik Rietdijk <notifications@github.com> wrote:
>FRidh commented on this pull request.
>
>
>
>> @@ -1345,6 +1345,15 @@ in
> 
>   duperemove = callPackage ../tools/filesystems/duperemove { };
> 
>+  dvc = callPackage ../applications/version-management/dvc { };
>+
>+  dvc-with-remotes = callPackage
>../applications/version-management/dvc {
>
>Is this variant needed? Users can override themselves, not.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/54530#pullrequestreview-307807549

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Review by worldofpeace — APPROVED** [review]
LGTM, I think the remotes variant is reasonable.
